Tools/Scripts/webkitpy/style/checkers/basexcconfig.py

```diff
@@ -0,0 +1,95 @@
+
+from webkitpy.common.memoized import memoized
+
+
+"""Enforces rules for Base.xcconfig files."""
+
+
+class BaseXcconfigChecker(object):
+    categories = set([
+        'basexcconfig/missing-wk_default-prefix',
+        'basexcconfig/missing-inherited',
+        'basexcconfig/overrides-common-base',
+    ])
+
+    default_vars = {
+        'GCC_OPTIMIZATION_LEVEL': 1,
+    }
+
+    def __init__(self, file_path, handle_style_error):
+        self._file_path = file_path
+        self._handle_style_error = handle_style_error
+
+    @memoized
+    def read_common_base_xcconfig_variables(self):
+        inherited_vars = {}
+        override_vars = {}
+
+        for line in open('Configurations/CommonBase.xcconfig'):
+            # Find lines containing assignments.
+            lhs, operator, rhs = map(str.strip, line.partition('='))
+
+            # Skip non-assignment lines, comments, and WK_ variables which are allowed to be overridden.
+            if operator != '=' or lhs.startswith('//') or lhs.startswith('WK_'):
+                continue
+
+            # Discard any setting condition.
+            name = lhs.partition('[')[0]
+
+            if '$(inherited)' in rhs:
+                inherited_vars[name] = 1
+            elif not self.default_vars.get(name):
+                override_vars[name] = 1
+
+        return inherited_vars, override_vars
+
+    def check(self, lines):
+        (inherited_vars, override_vars) = self.read_common_base_xcconfig_variables()
+
+        for line_number, line in enumerate(lines, start=1):
+            # Find lines containing assignments.
+            lhs, operator, rhs = map(str.strip, line.partition('='))
+
+            # Skip non-assignment lines, comments, and WK_ variables which are allowed to be overridden.
+            if operator != '=' or lhs.startswith('//') or lhs.startswith('WK_'):
+                continue
+
+            # Discard any setting condition.
+            name = lhs.partition('[')[0]
+
+            if self.default_vars.get(name):
+                self._handle_style_error(
+                    line_number, 'basexcconfig/missing-wk_default-prefix', 5,
+                    '{name} must be declared as WK_DEFAULT_{name} (see CommonBase.xcconfig)'.format(name=name))
+            elif inherited_vars.get(name):
+                inherited_name = '$(WK_COMMON_{name})'.format(name=name)
+                if '$(inherited)' not in rhs and inherited_name not in rhs:
+                    self._handle_style_error(
+                        line_number, 'basexcconfig/missing-inherited', 5,
+                        '{name} must include \'$(inherited)\' or \'$(WK_COMMON_{name})\' (see CommonBase.xcconfig)'.format(name=name))
+            elif override_vars.get(name):
+                self._handle_style_error(
+                    line_number, 'basexcconfig/overrides-common-base', 4,
+                    '{name} overrides the same variable in CommonBase.xcconfig'.format(name=name))
```

Comment:
At the very least, could we suggest a workaround in this error message? My fear is that just saying "you can't override this" will encourage people to ignore the style checker.

Reply:
I used `4` instead of `5` in `_handle_style_error()` to denote this isn't an error all the time.

I could change it to `3`, but I don't think it will be reported as an issue then.

Initially I think we should only move values into `CommonBase.xcconfig` that are intended to be the same across all projects.  If we need a way to ignore project-specific values, we can come up with that later.  (Or I could remove that check for now.)

Reply:
Keeping as-is is fine with me! We can live with us, and improve it as needed 🙂